src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,358 @@
+            {
+                return;
+            }
+
+            var vsShell = _serviceProvider.GetService(typeof(SVsShell)) as IVsShell;
+            var hr = vsShell.IsPackageInstalled(ReSharperPackageGuid, out int extensionEnabled);
+            if (ErrorHandler.Failed(hr))
+            {
+                FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
+                return;
+            }
+
+            var currentStatus = _workspace.Options.GetOption(KeybindingResetOptions.ReSharperStatus);
+
+            if (extensionEnabled == 0)
+            {
+                // If 0, the extension is either disabled in the extension manager, or not installed at all.
+                // If this is a change, update the status.
+                bool needsReset = _workspace.Options.GetOption(KeybindingResetOptions.NeedsReset);
+                if (currentStatus != ReSharperStatus.NotInstalledOrDisabled)
+                {
+                    // If we're going directly from Enabled->NotInstalled, we need to reset keybindings. Otherwise, the previous value of NeedsReset
+                    // is correct.
+                    var changedOptions = _workspace.Options;
+                    if (currentStatus == ReSharperStatus.Enabled)
+                    {
+                        needsReset = true;
+                        changedOptions = changedOptions.WithChangedOption(KeybindingResetOptions.NeedsReset, true);
+                    }
+
+                    changedOptions = _workspace.Options.WithChangedOption(KeybindingResetOptions.ReSharperStatus, ReSharperStatus.NotInstalledOrDisabled);
+
+                    _workspace.Options = changedOptions;
+                }
+
+                if (needsReset)
+                {
+                    ShowGoldBar();
+                }
+            }
+            else
+            {
+                UpdateReSharperEnableStatus();
+
+                // We need to monitor for suspend/resume commands, so create and install the command target.
+                var priorityCommandTargetRegistrar = _serviceProvider.GetService(typeof(SVsRegisterPriorityCommandTarget)) as IVsRegisterPriorityCommandTarget;
+                hr = priorityCommandTargetRegistrar.RegisterPriorityCommandTarget(
+                    dwReserved: 0 /* from docs must be 0 */,
+                    pCmdTrgt: this,
+                    pdwCookie: out _priorityCommandTargetCookie);
+
+                // Initialize the OleComponent to listen for modal changes (which will tell us when Tools->Options is closed)
+                _oleComponent = OleComponent.CreateHostedComponent("Keybinding Reset Detector");
+                _oleComponent.ModalStateChanged += OnModalStateChanged;
+
+                if (ErrorHandler.Failed(hr))
+                {
+                    FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
```

Comment:
shouldn't we check this before doing _oleComment and event registration? otherwise, who will clean those up when it failed here?